### ISSUE TITLE
Drag Digitakt samples to track on active pattern

### ISSUE THREAD

**Issue by szszoke** [dagargo/elektroid#136]
Apparently this is possible with Loopcloud. Would this be something in scope for Elektroid?

**Comment by dagargo** [comment]
It's been a while but support for uploading samples to tracks has been finally implemented. Also, RAM can be accessed and it's possible to upload and download samples and delete sample slots.

For anyone interested, check the `feat-digitakt-upload-to-track` branch. Although experimental, it seems to work well. Notice that the shown tracks are the ones in the currently selected pattern.

Keep in mind that this has only been tested with a Digitakt, which is the only Elektron machine to have such features.

**Comment by szszoke** [comment]
I have sold my Digitakt since opening this issue, so the best I can do is tp test it on the Digitakt 2.

Do you have any suggestions for that?

**Comment by szszoke** [comment]
I had some time to look at the segfault with a debugger. `sample->info` is a null pointer in `sample.c:sample_get_memfile_from_sample`, which then gets stored in `sample_info`. When you try to use `sample_info->channels` then it segfaults because the null pointer.

Edit: downloading samples via the `Samples` file system still works, so I don't know what is going.

**Comment by szszoke** [comment]
> I've pushed some commits and the segfault should be addressed as it seemed to be caused by a non initialized string and in the case of a `NULL` value it was continuing when it should have stopped. Also, some issues for handling the `NULL` at the graphical level have been addressed too.
> 
> It seems the issue is that the path to which the RAM is pointing to is `NULL`.
> 
> This is what you should get from the RAM when starting
> 
> ```
> $ elektroid-cli elektron:ram:ls 2:/ | head -n 2
>             001 
>             002 
> $ elektroid-cli elektron:ram:ul kick.wav 2:/1
> $ elektroid-cli elektron:ram:ul snare.wav 2:/2
> $ elektroid-cli elektron:ram:ls 2:/ | head -n 2
> F   28.19KiB 001 /incoming/kick                   [ used ]
> F   18.81KiB 002 /incoming/snare                  [ used ]
> ```
> 
> By the way, I've noticed that when a file is deleted the RAM slot still shows the path as if it were still loaded. This happens when the sample is delete from both the machine or Elektroid.

No change to the segfault. The same line still segfaults. `sample->info` is still a null pointer.

This is what I get from `elektroid-cli elektron:ram:ls 5:/ | head -n 2`:

```
F   372.2KiB 001 /factory/1-Drum hits/1-Bass drums/BD Hard/BD Palak [ used ]
F   41.66KiB 002 /factory/1-Drum hits/2-Snares/SD Snappy/SD Stonk [ used ]
```

I don't have time at the moment to familiarize myself with the Elektroid source code, and to experiment, so that will have to wait.

**Comment by dagargo** [comment]
Definitely, downloading a factory sample was causing a SIGSEGV. The issue was very old and was not happening because the factory samples are not even shown in Elektroid.

The problem was a bad check over the status of the open-for-reading operation over a sample path. Now, it should fail but should not crash.

Fixed in c1e661c0b15f80260f9824039510a21f60a9a261 and pushed.

**Comment by szszoke** [comment]
Can confirm. Downloading the factory sample now fails with a `Terminated with errors` message, and downloading a sample that I uploaded myself works.

Interestingly, uploading to a sample to a track or RAM fails now with the same error.

**Comment by dagargo** [comment]
A couple of notes.

As the code is merged, this is considered completed but I'm leaving this open because the integration with the Digitakt II is still unclear. I'm removing this from the 3.3 milestone too.

Regarding my previous comment about applying the same approach to the KORG volca sample 2, it is not possible because the volcas don't store the tempo at a pattern level and it is not possible to retrieve the _global_ tempo either AFAIK.
